lib/srv/alpnproxy/conn_upgrade.go

```diff
@@ -0,0 +1,153 @@
+// way to our Proxy Service.
+//
+// In those cases, the Teleport client should make a HTTP "upgrade" call to the
+// Proxy Service to establish a tunnel for the originally planned traffic to
+// preserve the ALPN and SNI information.
+func IsALPNConnUpgradeRequired(addr string, insecure bool) bool {
+	netDialer := &net.Dialer{
+		Timeout: defaults.DefaultDialTimeout,
+	}
+	tlsConfig := &tls.Config{
+		NextProtos:         []string{string(common.ProtocolReverseTunnel)},
+		InsecureSkipVerify: insecure,
+	}
+	testConn, err := tls.DialWithDialer(netDialer, "tcp", addr, tlsConfig)
+	if err != nil {
+		// If dialing TLS fails for any reason, we assume connection upgrade is
+		// not required so it will fallback to original connection method.
+		//
+		// This includes handshake failures where both peers support ALPN but
+		// no common protocol is getting negotiated. We may have to revisit
+		// this situation or make it configurable if we have to get through a
+		// middleman with this behavior. For now, we are only interested in the
+		// case where the middleman does not support ALPN.
+		logrus.Infof("ALPN connection upgrade test failed for %q: %v.", addr, err)
+		return false
+	}
+	defer testConn.Close()
+
+	// Upgrade required when ALPN is not supported on the remote side so
+	// NegotiatedProtocol comes back as empty.
+	result := testConn.ConnectionState().NegotiatedProtocol == ""
+	logrus.Debugf("ALPN connection upgrade required for %q: %v.", addr, result)
+	return result
+}
+
+// alpnConnUpgradeDialer makes an "HTTP" upgrade call to the Proxy Service then
+// tunnels the connection with this connection upgrade.
+type alpnConnUpgradeDialer struct {
+	netDialer *net.Dialer
+	insecure  bool
+}
+
+// newALPNConnUpgradeDialer creates a new alpnConnUpgradeDialer.
+func newALPNConnUpgradeDialer(keepAlivePeriod, dialTimeout time.Duration, insecure bool) ContextDialer {
+	return &alpnConnUpgradeDialer{
+		insecure: insecure,
+		netDialer: &net.Dialer{
+			KeepAlive: keepAlivePeriod,
+			Timeout:   dialTimeout,
+		},
+	}
+}
+
+// DialContext implements ContextDialer
+func (d alpnConnUpgradeDialer) DialContext(ctx context.Context, network, addr string) (net.Conn, error) {
+	logrus.Debugf("ALPN connection upgrade for %v.", addr)
+
+	tlsConn, err := tls.DialWithDialer(d.netDialer, network, addr, &tls.Config{
+		InsecureSkipVerify: d.insecure,
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	err = upgradeConnThroughWebAPI(tlsConn, url.URL{
+		Host:   addr,
+		Scheme: "https",
+		Path:   connectionUpgradeWebAPI,
+	})
+	if err != nil {
+		defer tlsConn.Close()
+		return nil, trace.Wrap(err)
+	}
+	return tlsConn, nil
+}
+
+func upgradeConnThroughWebAPI(conn net.Conn, api url.URL) error {
+	req, err := http.NewRequest(http.MethodGet, api.String(), nil)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// For now, only "alpn" is supported.
+	req.Header.Add("Upgrade", "alpn")
```

Comment:
nit: Move the `"alpn"` value to a constant, so we can use it here and at `lib/web/conn_upgrade.go`. Maybe placing a new file at `alpnproxy/common/upgrade.go` could be imported in both parts currently used.

Reply:
I feel these constant belongs more to the web server instead of alpn lib. Though I didn't want alpn lib depends on web or vice versa. I just moved these constants to `teleport/constants.go` now. 

For `h.GET("/webapi/connectionupgrade", httplib.MakeHandler(h.connectionUpgrade))` I would keep it this way instead of using constants as it aligns with other declarations and very clear what the endpoint is.